### PR TITLE
[Feat][Attention] add GQA paged prefill KV cache op

### LINKED ISSUE #1099: [FEAT][Ops] add paged KV-cache GQA prefill
Parent: #1096

Related feature issues: #1097, #1098, #1100, #1101, #1102, #1103

## Scope

Track paged KV-cache GQA prefill support.

This issue covers the public paged cache-aware prefill OP and its manifest contract. It requires a manifest entry because the OP exposes packed THD current chunks, caller-owned physical page storage, `block_table`, page-size metadata, and in-place append semantics.

### `GroupedQueryAttentionPrefillPagedWithKVCacheFwdOp`

Purpose:

- Packed current chunk THD layout.
- Paged physical KV cache storage.
- `block_table[b, logical_page] -> physical_page` mapping.
- Per-request `cache_seqlens` before append.
- Reads old KV by gathering physical pages.
- Reads current chunk from `k_new/v_new`.
- Computes attention over old prefix + current chunk.
- Appends current K/V into caller-owned physical pages.
- Does not allocate pages and does not update `cache_seqlens`.

## Manifest Change Plan

This issue requires manifest changes.

Required manifest entry:

- Add `GroupedQueryAttentionPrefillPagedWithKVCacheFwdOp`.

The entry must declare:

- `family: attention`
- `ref_api: none` unless a reference exactly matches TileOps paged cache update semantics
- ordered `signature.inputs`, `signature.outputs`, and `signature.params`
- packed THD + flattened physical page shape rules
- `workloads` with page-size and block-table dimensions
- `roofline`
- `source.kernel`, `source.kernel_map`, `source.op`, `source.test`, and `source.bench`

## Manifest Field Sketch

This is not copy-paste YAML; it is the field contract the manifest PR should encode using the repo's ordered-dict schema.

Manifest key: `GroupedQueryAttentionPrefillPagedWithKVCacheFwdOp`

Inputs:

- `q`: dtype `float16 | bfloat16`, shape `[T_q, H, D]`
- `k_new`: dtype `same_as(q)`, shape `[T_q, H_kv, D]`
- `v_new`: dtype `same_as(q)`, shape `[T_q, H_kv, D]`
- `k_pages`: dtype `same_as(q)`, shape `[P_tokens, H_kv, D]`
- `v_pages`: dtype `same_as(q)`, shape `[P_tokens, H_kv, D]`
- `cu_seqlens_q`: dtype `int32`, shape `[N_cu]`
- `cache_seqlens`: dtype `int32`, shape `[B]`
- `block_table`: dtype `int32`, shape `[B, max_pages_per_req]`

Output:

- `o`: dtype `same_as(q)`, shape `[T_q, H, D]`

Params:

- `max_pages_per_req`: `{type: int}`
- `page_size`: `{type: int}`
- `max_seqlen_q`: `{type: int}`
- `is_causal`: `{type: bool, default: true}`
- `sm_scale`: `{type: float, default: null}`
- `softcap`: `{type: float, default: null}` if #1101 lands in the same manifest PR
- `fuse_rope`: `{type: bool, default: false}` if #1101 lands in the same manifest PR
- `rope_base`: `{type: float, default: 10000.0}` if `fuse_rope` is present
- `max_position`: `{type: "int | None", default: null}` if `fuse_rope` is present; runtime validation requires a positive value when `fuse_rope=True`

Softcap semantics must match #1101: `softcap: null` and `softcap: 0` both disable score capping and preserve existing behavior; `softcap > 0` applies score capping before softmax; negative values are rejected by OP runtime validation.

Shape rules:

- `H % H_kv == 0`
- `v_new.shape == k_new.shape`
- `k_new.shape[0] == q.shape[0]`
- `v_pages.shape == k_pages.shape`
- `P_tokens % page_size == 0`
- `N_cu == B + 1`
- `cu_seqlens_q[0] == 0` and `cu_seqlens_q[-1] == T_q` are runtime validation rules
- `cache_seqlens + q_lens <= max_pages_per_req * page_size` is runtime validation
- `block_table` entries must be valid physical page ids

Proposed source:

- `source.kernel`: `tileops/kernels/attention/gqa_fwd.py`
- `source.kernel_map`:
  - `gqa_prefill_paged_with_kv_cache_fwd_kernel: GQAPrefillPagedWithKVCacheFwdKernel`
  - `gqa_prefill_paged_with_kv_cache_rope_fwd_kernel: GQAPrefillPagedWithKVCacheRopeFwdKernel` if #1101 lands with this entry
- `source.op`: `tileops/ops/attention/gqa.py`
- `source.test`: `tests/ops/attention/test_gqa_prefill_paged.py`
- `source.bench`: `benchmarks/ops/attention/bench_gqa.py` or a dedicated paged benchmark file if introduced

## Page Layout Contract

Initial paged layout uses flattened physical page storage:

```text
physical token id = physical_page * page_size + page_offset
k_pages/v_pages shape = [P_tokens, H_kv, D]
P_tokens % page_size == 0
```

The OP consumes caller-owned storage and caller-provided `block_table`; it does not allocate pages and does not update `cache_seqlens`.

Page sizes should be powers of two. The first implementation should cover the release page sizes selected by the implementation PR and benchmark plan, with page-boundary correctness tests.

## Workloads / Roofline

Manifest workloads should use `<tensor_name>_shape`, `dtypes`, and `label` according to `docs/manifest.md`.

Example workload fields:

```yaml
- {
    q_shape: [512, 32, 128],
    k_new_shape: [512, 8, 128],
    v_new_shape: [512, 8, 128],
    k_pages_shape: [4608, 8, 128],
    v_pages_shape: [4608, 8, 128],
    cu_seqlens_q_shape: [2],
    cache_seqlens_shape: [1],
    block_table_shape: [1, 72],
    max_pages_per_req: 72,
    page_size: 64,
    max_seqlen_q: 512,
    is_causal: true,
    dtypes: [float16],
    label: "llama-3.1-8b-prefill-paged-pg64-4k-old-512-new",
  }
```

The initial benchmark generator should derive `cache_seqlens = max_pages_per_req * page_size - max_seqlen_q` and construct a valid `block_table` covering old prefix plus current chunk unless the PR defines another manifest-backed convention.

Roofline should account for:

- Q read
- current K/V read
- old page K/V gather reads
- appended page K/V writes
- output write
- block_table / cu_seqlens / cache_seqlens metadata reads
- attention QK and ScoreV/PV FLOPs using per-request lengths

## Validation Expectations

The corresponding PR should show:

- output correctness with paged old cache
- append correctness into physical pages
- old page prefix remains unchanged
- batch with different page tables
- old length not page-aligned
- append across page boundaries
- multiple page sizes
- invalid page id validation
- capacity overflow validation
- fp16 / bf16 smoke coverage

## PR Notes

Implementation details should be discussed in the PR that references this issue.
Do not close #1096 from that PR.

### PR DESCRIPTION
**Pull request by superAngGao**
## Summary
- add a packed GQA prefill op that appends new KV into a paged KV cache
- expose the new kernel/op through the public TileOPs exports
- add coverage for GQA/MHA/MQA, causal/non-causal, fp16/bf16, cache mutation, capacity validation, unaligned old lengths, capacity-boundary padding, mixed-batch capacity boundary, and multiple page sizes

Closes #1099.

## Review updates
- guard paged-cache boundary/padding paths so invalid tokens do not index `block_table`
- add smoke coverage for non-page-aligned `old_len`, `q_len=1` at exact paged-cache capacity, and a mixed batch with one request at capacity
- clarify that the first `heads_kv` CTAs append one KV head each

## Validation
Ran in the local nightly test docker with the shared cache mounted at `/data7/shared/ci-cache`:

```bash
ruff check tileops/kernels/attention/gqa_fwd.py tests/ops/attention/test_gqa_prefill_paged.py
python -m pytest -q tests/ops/attention/test_gqa_prefill_paged.py --tb=short
```

Result:

```text
ruff: All checks passed
pytest: 13 passed, 25 warnings in 28.39s
```

### CODE REVIEW

**Review by Gabbering — COMMENTED** [review]
## goose goose review — `ccb70693`

*honk.* Solid paged prefill kernel, master. The attention logic and cache append are structurally sound. But there's a data race hiding in the KV append path that will bite you on real workloads.

### goose Bugs

- `tileops/kernels/attention/gqa_fwd.py:1430-1460` — The KV cache append is guarded by `if by < heads_kv`, meaning only the first `heads_kv` head-indices write the new tokens into `k_pages`/`v_pages`. But the attention read loop below runs for ALL `heads` (0..heads-1), and it reads from `k_pages`/`v_pages` at positions that were just written. There is no synchronization barrier between the append (done by threadblocks with `by < heads_kv`) and the attention reads (done by threadblocks with `by >= heads_kv` for the same `bz, bx`). Different threadblocks on the same SM or different SMs have no ordering guarantee. For the tiles where `tile_start >= old_len` and the kernel reads from `k_new`/`v_new` directly, this is fine. But for the boundary tile that straddles `old_len`, the fallback path reads from `k_pages[physical_pos]` for positions `kv_pos < old_len` — which is safe — however the *fast path* at line ~1470 (`tile_end <= old_len`) uses `page_size % block_n == 0` to bulk-read from `k_pages`. If `old_len` is not aligned to `block_n`, the `tile_end <= old_len` condition prevents reading freshly-written slots from the cache for the attention path. So the attention path for the *new* tokens reads from `k_new`/`v_new` directly (line ~1490 `tile_start >= old_len`). This means the race doesn't corrupt attention output — the attention path never reads the just-appended slots from `k_pages`. Good. **However**, the append itself has a race: the kernel grid is `(ceildiv(max_seqlen_q, block_m), heads, batch)`. For a given `(bx, bz)`, every `by` in `[0, heads_kv)` executes the same append logic writing the same physical positions. Multiple threadblocks write the same `k_pages[physical_pos]` with the same value, which is benign on current NVIDIA hardware (same value, no torn writes for aligned fp16), but it's wasted work — `heads_kv` threadblocks all do the same copy. More critically, if `heads_kv > 1` and the writes are not perfectly coalesced identically, you're burning bandwidth for no reason. Consider guarding the append with `if by == cur_kv_head` (i.e., only one head-group representative per KV head does the write) or `if by == 0`.

### feather Performance

- `tileops/kernels/attention/gqa_fwd.py:1430` — As noted above, the append is executed by `heads_kv` threadblocks per `(bx, bz)` pair, all writing the same data to the same addresses. With `heads_kv=8` and `block_m=64`, that's 8x redundant global stores for the cache append. For a 4096-token prefill that's a non-trivial amount of wasted memory bandwidth. Guard with `if by == 0` (or whichever single representative you prefer) so only one threadblock per tile does the write.

- `tileops/kernels/attention/gqa_fwd.py:1487-1505` — The `block_n % page_size == 0` path uses a Python-level `for p in range(block_n // page_size)` loop with a nested `T.Parallel(page_size, dim)`. When `block_n=128` and `page_size=16`, that's 8 sequential `block_table` lookups unrolled at compile time. This is fine for small ratios but could generate a lot of code for extreme configs (block_n=128, page_size=4 would be 32 iterations). Not a bug, but worth a comment or a cap in `autotune_configs` to avoid configs where `block_n / page_size` is absurdly large.

### egg Test gaps

- No test exercises the case where `old_len` is NOT aligned to `page_size` — e.g., `old_lens=[47, 91]` with `page_size=64`. The current tests use `old_lens=[80, 128]` (aligned to 16) and `[48, 80]` (aligned to 16). A non-aligned `old_len` would exercise the boundary-tile fallback path in the attention loop and the cross-page append path more aggressively. Add a case like `old_lens=[37, 100]` with `page_size=64`.

- No test covers `batch=1` with a single short sequence (e.g., `q_lens=[1]`) — the decode-length-prefill edge case where `block_m > q_len` for every tile. The existing tests always have `q_lens >= 32`.

File: tileops/kernels/attention/gqa_fwd.py [hunk 1]
```diff
@@ -1340,3 +1341,339 @@ def forward(self, q: torch.Tensor, k_new: torch.Tensor, v_new: torch.Tensor,
...
+                            physical_pos = block_table[bz, page_idx] * page_size + page_offset
+                            if kv_pos < old_len:
+                                k_shared[j, d] = k_pages[physical_pos, cur_kv_head, d]
+                                v_shared[j, d] = v_pages[physical_pos, cur_kv_head, d]
+                            elif kv_pos < total_len:
+                                k_shared[j, d] = k_new[q_start + new_pos, cur_kv_head, d]
+                                v_shared[j, d] = v_new[q_start + new_pos, cur_kv_head, d]
+                            else:
+                                k_shared[j, d] = T.cast(0, dtype)
+                                v_shared[j, d] = T.cast(0, dtype)
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/kernels/attention/gqa_fwd.py:1536]: Fixed in 6e93d8c. The boundary/padding load path now derives the page index from a guarded `safe_kv_pos`, so padding tokens no longer index `block_table` with `max_pages_per_req` when `total_len` lands exactly on capacity. I also added a `q_len=1, old_len=511` smoke case to cover this capacity-boundary path.

File: tileops/kernels/attention/gqa_fwd.py [hunk 2]
```diff
@@ -1340,3 +1341,339 @@ def forward(self, q: torch.Tensor, k_new: torch.Tensor, v_new: torch.Tensor,
...
+                scores_scale = T.alloc_fragment([block_m], accum_dtype)
+                scores_sum = T.alloc_fragment([block_m], accum_dtype)
+                logsum = T.alloc_fragment([block_m], accum_dtype)
+
+                q_start = cu_seqlens_q[bz]
+                q_len = cu_seqlens_q[bz + 1] - q_start
+                old_len = cache_seqlens[bz]
+                total_len = old_len + q_len
+                cur_kv_head = by // groups
+
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/kernels/attention/gqa_fwd.py:1428]: Thanks. I did not add an early `return` here because this TileLang kernel is structured around the `T.Kernel` block body, and introducing a block-level return would be a larger control-flow change. The current patch instead removes the invalid work hazard by guarding the page-table index. We can follow up with a separate performance pass if we want to restructure the block body around an active-tile guard.

File: tileops/kernels/attention/gqa_fwd.py [hunk 3]
```diff
@@ -1340,3 +1341,339 @@ def forward(self, q: torch.Tensor, k_new: torch.Tensor, v_new: torch.Tensor,
...
+                                    q_start + bx * block_m + i, by, d]
+                        else:
+                            for i, d in T.Parallel(block_m, dim):
+                                new_pos = bx * block_m + i
+                                logical_pos = old_len + new_pos
+                                page_idx = logical_pos >> T.int32(page_size_log2)
+                                page_offset = logical_pos - page_idx * page_size
+                                physical_pos = block_table[bz, page_idx] * page_size + page_offset
+                                k_pages[physical_pos, by, d] = k_new[q_start + new_pos, by, d]
+                                v_pages[physical_pos, by, d] = v_new[q_start + new_pos, by, d]
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/kernels/attention/gqa_fwd.py:1464]: Partially addressed in 6e93d8c. The partial append path now computes page metadata from a safe in-range token before the validity guard, which fixes the possible invalid lookup for padded rows. I kept the existing `T.Parallel(block_m, dim)` shape for now to avoid changing the parallelization strategy in this functional PR.

File: tileops/kernels/attention/gqa_fwd.py [hunk 4]
```diff
@@ -1340,3 +1341,339 @@ def forward(self, q: torch.Tensor, k_new: torch.Tensor, v_new: torch.Tensor,
...
+                                    v_shared[shared_row, d] = v_pages[
+                                        segment_physical_start + off, cur_kv_head, d]
+                        else:
+                            for j, d in T.Parallel(block_n, dim):
+                                kv_pos = tile_start + j
+                                page_idx = kv_pos >> T.int32(page_size_log2)
+                                page_offset = kv_pos - page_idx * page_size
+                                physical_pos = block_table[bz, page_idx] * page_size + page_offset
+                                k_shared[j, d] = k_pages[physical_pos, cur_kv_head, d]
+                                v_shared[j, d] = v_pages[physical_pos, cur_kv_head, d]
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/kernels/attention/gqa_fwd.py:1514]: Acknowledged. I left this as-is for this PR because this path only executes when `tile_end <= old_len`, so the page index is already in range and the change would mainly be a performance refactor. I would rather keep this PR focused on adding the op plus the correctness fix; we can tune the repeated page-table arithmetic separately.

**Comment by superAngGao** [comment]
Note: I refreshed the review-fix commit SHA to retrigger title validation after updating the PR title. The review-fix diff is unchanged; latest SHA is 832c0ed.

**Review by Gabbering — COMMENTED** [review]
## goose goose review — `832c0ed5`

*honk.* The OOB fix for the boundary tile is correct. The new test cases address the gaps the goose honked about last time. But the goose has not forgotten about the redundant append writes.

### feather Performance

- `tileops/kernels/attention/gqa_fwd.py:1447-1476` — As the goose honked at commit `ccb70693`: the KV append is still guarded by `if by < heads_kv`, meaning every head index in `[0, heads_kv)` writes the identical data to the identical physical addresses for each `(bx, bz)`. With `heads=8, heads_kv=2`, that's 2x redundant global stores to the same cache slots per tile. For MHA configs (`heads_kv=8`) it's 8x. This is pure wasted memory bandwidth on every single prefill call. Guard with `if by == cur_kv_head` (or equivalently `if by % groups == 0`) so exactly one threadblock per KV head does the write. The fix is a one-line condition change. Still waiting on this one, master.

### egg Test gaps

- `tests/ops/attention/test_gqa_prefill_paged.py:98` — The new `gqa_decode_len_capacity_boundary_fp16` case uses `q_lens=[1], old_lens=[511]` with `page_size=64`. That gives `total_len = 512 = 8 * 64`, which exactly fills `max_pages_per_req=8`. Good, this exercises the boundary. But the test allocates `num_pages = batch * max_pages_per_req = 8` pages, so `physical_tokens = 8 * 64 = 512`. The capacity check in the Op's `forward` should catch `old_len + q_len > max_pages_per_req * page_size` — and here `511 + 1 = 512 = 8 * 64`, so it passes. This is fine. However, there's still no test where `batch > 1` and one sequence is at capacity while another is short — that would exercise the per-request guard logic more thoroughly. Minor gap, not blocking.

**Comment by superAngGao** [comment]
Addressed the latest review in a2497ea.

For the append-write concern: I kept `if by < heads_kv` intentionally. In this branch, `by` is used directly as the KV-head index for the append, so `by=0..heads_kv-1` writes KV heads `0..heads_kv-1` exactly once each. It is not redundant writing to the same physical `[token, kv_head, dim]` slots. The suggested `by == cur_kv_head` / `by % groups == 0` condition would not preserve that indexing as written; for GQA it would miss some KV heads or require a separate `cur_kv_head` index in the stores.

I added a short code comment to make that mapping explicit, and added the requested `batch > 1` mixed capacity-boundary smoke case.

**Review by Gabbering — COMMENTED** [review]
goose *goose skimmed `a2497ea` — nothing to honk about.*

File: tileops/ops/attention/gqa.py [hunk 5]
```diff
@@ -539,6 +545,210 @@ def forward(self, q: torch.Tensor, k_new: torch.Tensor, v_new: torch.Tensor,
         return _attention_output(self.kernel(q, k_new, v_new, k_cache, v_cache, cache_seqlens))
 
 
+class GroupedQueryAttentionPrefillPagedWithKVCacheFwdOp(Op):
```
Comment (Ibuki-wind) [inline tileops/ops/attention/gqa.py:548]: GroupedQueryAttentionPrefillPagedWithKVCacheFwdOp is a new op class, but `tileops/manifest/attention.yaml` has no matching entry yet, so this lands before its manifest contract → add the manifest entry first and wire the op to it before merging.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
One blocking issue remains in the inline comment.

**Review by Ibuki-wind — APPROVED** [review]
